README.md

```diff
@@ -22,50 +22,118 @@ although basic usage is shown below.
+  return res.send({result});
 }
 ```
 
+##### Client logic:
+
+Note that the `window.u2f` object is defined in the official [Javascript U2F API](https://github.com/google/u2f-ref-code), for which a polyfill is [available as an npm module](https://www.npmjs.com/package/u2f-api-polyfill).
+
+```javascript
+const registrationReq = ...  // Retrieve this from hitting the registration challenge endpoint
+const {appId, version, challenge} = registrationReq;
+
+window.u2f.register(appId, [{version, challenge}], [], (registrationRes) => {
+  // Send this registration response to the registration verification server endpoint
+});
+```
+
 ### User Authentication Flow
+
+##### Server endpoint (authentication challenge):
+
 ```javascript
-// 1. Check that the user is logged in using password authentication (or some other way).
+const u2f = require('u2f');
+
+function handler(req, res) {
+  // 1. Check that the user is logged in.
 
-// 2. (On server) generate user sign request using keyHandle from user account.
-var req = u2f.request("<appId>", user.keyHandle);
-session.authRequest = req;
+  // 2. Fetch the user's key handle from the server-side datastore. This field should have been
+  // saved after the registration procedure.
+  const keyHandle = ...
 
-// 3. (On client) use U2F API to request signature.
-// u2f.sign([req], function(res) { /* send res back to server */ });
+  // 3. Generate an authentication request and save it in the session. Use the same app ID that
+  // was used in registration!
+  const authReq = u2f.request('<appId>', keyHandle);
+  req.session.authReq = authReq;
 
-// 4. (On server) check signature using publicKey from user account.
-var checkres = u2f.checkSignature(session.authRequest, res, user.publicKey);
+  // 4. Send the authentication request to the client, who will use the Javascript U2F API to sign
+  // the authentication request, and send it back to the server for verification.
+  return res.send(authReq);
+}
+```
 
-if (checkres.successful) {
+##### Server endpoint (authentication verification):
+
+```javascript
+const u2f = require('u2f');
+
+function handler(req, res) {
+  // 4. Verify the authentication response from the client against the authentication request saved
+  // in the server-side session.
+  const result = u2f.checkSignature(req.session.authReq, req.body.authRes);
+
+  if (result.successful) {
+    // Success!
     // User is authenticated.
-} else {
-    // checkres.errorMessage will contain error text.
+    return res.sendStatus(200);
+  }
+
+  // result.errorMessage is defined with an English-language description of the error.
+  return res.send({result});
 }
 ```
 
+##### Client logic:
+
+```javascript
+const authReq = ...;  // Retrieve this from hitting the authentication challenge endpoint
+const {appId, challenge, version, keyHandle} = authReq;
+
+window.u2f.sign(appId, challenge, [{version, keyHandle}], (authRes) => {
```

Comment:
The input parameters to `u2f.sign` are fixed here relative to what's already in the README. Reference: [official Javascript U2F client API implementation](https://demo.yubico.com/js/u2f-api.js) and [FIDO API spec](https://fidoalliance.org/specs/fido-u2f-v1.0-nfc-bt-amendment-20150514/fido-u2f-javascript-api.html)

Relevant snippet:

```
/**
 * Dispatches an array of sign requests to available U2F tokens.
 * If the JS API version supported by the extension is unknown, it first sends a
 * message to the extension to find out the supported API version and then it sends
 * the sign request.
 * @param {string=} appId
 * @param {string=} challenge
 * @param {Array<u2f.RegisteredKey>} registeredKeys
 * @param {function((u2f.Error|u2f.SignResponse))} callback
 * @param {number=} opt_timeoutSeconds
 */
u2f.sign = function(appId, challenge, registeredKeys, callback, opt_timeoutSeconds) {
```